### PR TITLE
Update World Prospector App

### PR DESCRIPTION
**Pull request by Caedis**
## What
Updates the world prospector app with a couple of minor changes.

* Fixes the battery drain when an AR app is "open", but is not in either hand
* Highlight color  is now based off the OreDictUnifier material if possible
* Highlight color is now auto selected for JEI phantom drags
* Clearing the block now also clears the highlight color
* Rendering pauses when game is paused

## Additional Information
[Preview of changes](https://cdn.discordapp.com/attachments/904846945095385138/1088798826309099651/screencast-1679659942.mp4)
[Discord link](https://discord.com/channels/701354865217110096/904846945095385138/1088798827206672444)

### CODE REVIEW

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 1]
```diff
@@ -119,6 +181,33 @@ public void handleClientAction(int id, PacketBuffer buffer) {
...
+                colors[index] = ms.material.getMaterialRGB();
+            } else {
+                colors[index] = block.getStateFromMeta(copy.getMetadata()).getMaterial().getMaterialMapColor().colorValue;
+            }
+            if (colors[index] == 0) {
+                colors[index] = block.hashCode();
+            }
+            colors[index] = colors[index] | 0xff000000;
+            buttonWidget.setFill(colors[index]);
+        } else if (stack == ItemStack.EMPTY) {
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:204]:
```suggestion
        } else if (stack.isEmpty()) {
```

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 2]
```diff
@@ -95,6 +143,20 @@ public boolean mouseClicked(int mouseX, int mouseY, int button) {
...
+                    else if (id == 0) {
+                        updateBlockSelectionAndColor(ItemStack.EMPTY, index, buttonWidget);
+                    }
+                    else if (id == 1) {
+                        PacketBuffer tmpBuffer = new PacketBuffer(buffer.copy());
+                        try {
+                            ItemStack stack = tmpBuffer.readItemStack();
+                            updateBlockSelectionAndColor(stack, index, buttonWidget);
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:156]: Would we prefer to log an error here, fallback to updating the selection to ItemStack.EMPTY, and then move on?

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 3]
```diff
@@ -95,6 +143,20 @@ public boolean mouseClicked(int mouseX, int mouseY, int button) {
                 public void handleClientAction(int id, PacketBuffer buffer) {
                     if (id == -1) {
                         selectReference(index, buttonWidget);
+                    }
+                    else if (id == 0) {
+                        updateBlockSelectionAndColor(ItemStack.EMPTY, index, buttonWidget);
+                    }
+                    else if (id == 1) {
+                        PacketBuffer tmpBuffer = new PacketBuffer(buffer.copy());
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:151]: Why does the PacketBuffer need to be copied here?

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 4]
```diff
@@ -81,12 +93,48 @@ public AbstractApplication initApp() {
...
+                        @Nonnull
+                        @Override
+                        public Rectangle getArea() {
+                            return rectangle;
+                        }
+
+                        @Override
+                        public void accept(@Nonnull Object ingredient) {
+                            if (ingredient instanceof ItemStack) {
+                                int mouseButton = Mouse.getEventButton();
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:113]: Is `Mouse` safe to access here? Does client/server sidedness need to be somehow checked first?

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 5]
```diff
@@ -81,12 +93,48 @@ public AbstractApplication initApp() {
...
             RectButtonWidget buttonWidget = new RectButtonWidget(x + i * 24, y + 18, 18, 18, 1);
             addWidget(new PhantomSlotWidget(handlers[i], 0, x + i * 24, y) {
+
+                @Override
+                public List<IGhostIngredientHandler.Target<?>> getPhantomTargets(Object ingredient) {
+                    if (!(ingredient instanceof ItemStack)) {
+                        return Collections.emptyList();
+                    }
+                    Rectangle rectangle = toRectangleBox();
+                    return Lists.newArrayList(new IGhostIngredientHandler.Target<Object>() {
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:103]: Do you know if this list gets appended to at all? If not, you could use `Collections.singletonList(etc)` instead.

File: src/main/java/gregtech/common/items/behaviors/TerminalBehaviour.java [hunk 6]
```diff
@@ -60,6 +60,14 @@ public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player
     public void onUpdate(ItemStack itemStack, Entity entity) {
         NBTTagCompound tabletNBT = itemStack.getOrCreateSubCompound("terminal");
         if (entity.ticksExisted % 20 == 0 && tabletNBT.hasKey("_ar")) {
+
+            if (entity instanceof EntityPlayer){
+                EntityPlayer ep = (EntityPlayer) entity;
+                if (!(ep.getHeldItem(EnumHand.MAIN_HAND) == itemStack || ep.getHeldItem(EnumHand.OFF_HAND) == itemStack)){
+                    return;
+                }
+            }
```
Comment (TechLord22) [inline src/main/java/gregtech/common/items/behaviors/TerminalBehaviour.java:69]:
```suggestion
            if (entity instanceof EntityLivingBase) {
                EntityLivingBase livingBase = (EntityLivingBase) entity;
                if (!livingBase.getHeldItemMainhand().isItemEqual(itemStack) && !livingBase.getHeldItemOffhand().isItemEqual(itemStack)) {
                    return;
                }
            }
```

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 7]
```diff
@@ -154,27 +243,10 @@ protected void hookDrawInBackground(int mouseX, int mouseY, float partialTicks,
...
-                    }
-                    colors[index] = colors[index] | 0xff000000;
-                    buttonWidget.setFill(colors[index]);
-                }
-
-            }).open();
+                stack -> stack.getItem() instanceof ItemBlock,
+                stack -> {
+                    updateBlockSelectionAndColor(stack, index, buttonWidget);
+                }).open();
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:249]:
```suggestion
                stack -> updateBlockSelectionAndColor(stack, index, buttonWidget)
        ).open();
```

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java (hunk 3 above)
Comment (Caedis) [reply src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:151]: `PacketBuffer#read` methods increase a read index which is how you can readInt, readItemStack, readBoolean back to back. I didnt see a way to reset it with a quick glance.

Turns out there is a way....in the clear method.
image

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java (hunk 2 above)
Comment (Caedis) [reply src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:156]: I took that logic from someone else where we read from a buffer.
Will add a logging message.

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java (hunk 4 above)
Comment (Caedis) [reply src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:113]: Dont think so, was copied from PhantomSlotWidget, just with 1 call added to the `accept` method

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java (hunk 3 above)
Comment (Caedis) [reply src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java]: Changed to marking the current reader index, then just resetting the read index

File: src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java [hunk 8]
```diff
@@ -319,6 +392,8 @@ private static void addCluster(BlockPos pos, Map<AxisAlignedBB, Set<BlockPos>> f
     @Override
     public void tickAR(EntityPlayer player) {
         World world = player.world;
+        if (Minecraft.getMinecraft().isGamePaused()) return;
```
Comment (TechLord22) [inline src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:395]: Does this need to check if `world.isRemote`, or is this method only called on the client?

Comment (Caedis) [reply src/main/java/gregtech/common/terminal/app/worldprospector/WorldProspectorARApp.java:395]: Client side only

**Review by TechLord22 — APPROVED** [review]

**Pull request merged**